### PR TITLE
Add ELF TLS support in new x64 backend.

### PR DESCRIPTION
**Pull request by cfallin**
Depends on #2538 and includes that commit in this PR.

This follows the implementation in the legacy x86 backend, including
hardcoded sequence that is compatible with what the linker expects. We
could potentially do better here, but it is likely not necessary.

Thanks to @bjorn3 for a bugfix to an earlier version of this.

### CODE REVIEW

**Comment by bjorn3** [comment]
cranelift/codegen/src/legalizer/tls.rs is empty

File: cranelift/codegen/src/isa/x86/enc_tables.rs [hunk 1]
```diff
@@ -1856,12 +1856,13 @@ fn convert_i64x2_imul(
...
 }
 
+/// Expand a `tls_value` instruction.
 fn expand_tls_value(
     inst: ir::Inst,
     func: &mut ir::Function,
     _cfg: &mut ControlFlowGraph,
     isa: &dyn TargetIsa,
-) {
+) -> bool {
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/x86/enc_tables.rs:1865]: Why this change?

File: cranelift/filetests/filetests/isa/x64/tls_elf.clif [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+gv0 = symbol colocated tls u1:0
+
+block0(v0: i32):
+    v1 = global_value.i64 gv0
+    return v1
+}
+
+; check:  pushq   %rbp
+; nextln: movq    %rsp, %rbp
+; nextln: elf_tls_get_addr User { namespace: 1, index: 0 }
```
Comment (bjorn3) [inline cranelift/filetests/filetests/isa/x64/tls_elf.clif:16]: Can you add a binemit test? The exact byte sequence is important.

File: cranelift/codegen/src/isa/x86/enc_tables.rs (hunk 1 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/x86/enc_tables.rs]: Ah, this (and `legalizer/tls.rs`) was leftover from an earlier version of the change; reverted to original. Thanks!

File: cranelift/filetests/filetests/isa/x64/tls_elf.clif (hunk 2 above)
Comment (cfallin) [reply cranelift/filetests/filetests/isa/x64/tls_elf.clif:16]: Unfortunately the `binemit` filetest backend doesn't support `MachInst` yet; I'll add an issue for that.

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 3]
```diff
@@ -4229,6 +4229,14 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             }
         }
 
+        Opcode::TlsValue => {
+            let dst = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
+            let (name, _, _) = ctx.symbol_value(insn).unwrap();
+            let symbol = name.clone();
+            ctx.emit(Inst::ElfTlsGetAddr { symbol });
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/x64/lower.rs:4236]: Can you add a `todo!()` for non elf tls?

**Review by bjorn3 — COMMENTED** [review]
LGTM with the `todo!()` added.

File: cranelift/codegen/src/isa/x64/lower.rs (hunk 3 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/x64/lower.rs]: Done!

**Review by bjorn3 — APPROVED** [review]

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 4]
```diff
@@ -2018,6 +2027,36 @@ fn x64_get_regs(inst: &Inst, collector: &mut RegUsageCollector) {
         | Inst::Fence { .. } => {
             // No registers are used.
         }
+
+        Inst::ElfTlsGetAddr { .. } => {
+            // All caller-saves are clobbered.
+            collector.add_def(Writable::from_reg(regs::rsi()));
```
Comment (bnjbvr) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:2033]: Would it make sense to use the return value of `ABIMachineSpec::get_regs_clobbered_by_call` here?

File: cranelift/codegen/src/isa/x64/inst/emit.rs [hunk 5]
```diff
@@ -2937,6 +2938,27 @@ pub(crate) fn emit(
         Inst::EpiloguePlaceholder => {
             // Generate no code.
         }
+
+        Inst::ElfTlsGetAddr { ref symbol } => {
```
Comment (bnjbvr) [inline cranelift/codegen/src/isa/x64/inst/emit.rs:2942]: Is there something meaningful we can add as a comment, explaining what the sequence does?
Also: can it be reframed as other Vcode::Inst instructions that would be recursively generated? This would act as documentation, and help understand and maintaining this sequence a bit better.

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Looks good on principle, I wonder if there are ways to minimize code duplication, see below. Thanks!

File: cranelift/codegen/src/isa/x64/inst/emit.rs (hunk 5 above)
Comment (bjorn3) [reply cranelift/codegen/src/isa/x64/inst/emit.rs:2942]: The thing is that it needs this exact byte sequence. Any tiny change will result in a linker error as the linker doesn't know how to rewrite it anymore. For the old backend I wrote a few comments that you could use here: https://github.com/bytecodealliance/wasmtime/blob/f579d088baac9df6bb852ecc19d8216fb94438a1/cranelift/codegen/meta/src/isa/x86/recipes.rs#L3356-L3377

Comment (cfallin) [reply cranelift/codegen/src/isa/x64/inst/emit.rs:2993]: Added comments similar to the original x86-backend ones, as well as a comment noting that the linker expects exactly these bytes.

File: cranelift/codegen/src/isa/x64/inst/mod.rs (hunk 4 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/x64/inst/mod.rs]: Ah yes, good idea! Done.

**Comment by cfallin** [comment]
Updated, thanks!

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 6]
```diff
@@ -2039,6 +2050,13 @@ fn x64_get_regs(inst: &Inst, collector: &mut RegUsageCollector) {
         | Inst::Fence { .. } => {
             // No registers are used.
         }
+
+        Inst::ElfTlsGetAddr { .. } => {
+            // All caller-saves are clobbered.
+            for reg in X64ABIMachineSpec::get_regs_clobbered_by_call(CallConv::SystemV) {
```
Comment (bnjbvr) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:2056]: Double-check me here: we can hardcore SystemV here because this instruction is specific for Elf, right? If that's not the case, could we get the call_conv from the caller; if that's the case, can you add a comment to that effect please?

**Review by bnjbvr — APPROVED** [review]
Thanks, lgtm!

File: cranelift/codegen/src/isa/x64/inst/mod.rs (hunk 6 above)
Comment (bjorn3) [reply cranelift/codegen/src/isa/x64/inst/mod.rs:2056]: Yes, this instruction is specific for ELF and must use the System-V call conv independently of the call conv of the function using TLS. Mach-O and PE use different TLS mechanisms.

Comment (cfallin) [reply cranelift/codegen/src/isa/x64/inst/mod.rs:2061]: Comment added!
